deepchem/models/logger.py

```diff
@@ -0,0 +1,64 @@
+from typing import Optional, Union, Dict, List
+import tensorflow as tf
+import torch
+import numpy as np
+from deepchem.models import Model
+
+numeric = Union[tf.Tensor, torch.Tensor, int, float, complex, np.number]
+tensor = Union[tf.Tensor, torch.Tensor]
+
+
+class Logger(object):
+
+  def __init__(self):
+    raise NotImplementedError
+
+  def __iter__(self):
+    return self
+
+  def setup(self, config: Dict):
+    raise NotImplementedError
+
+  def finish(self):
+    raise NotImplementedError
+
+  def log_batch(self,
+                loss: Dict,
+                step: int,
+                inputs: tensor,
+                labels: tensor,
+                location: Optional[str] = None,
+                model: Optional[Model] = None,
+                checkpoint_metric: Optional[str] = None,
+                checkpoint_metric_value: Optional[numeric] = None,
+                checkpoint_on_min: bool = True):
```

Comment:
I've been debating about this.  I don't like this API too much.  I know I'm the one who suggested the logger's checkpointing mechanism should be independent of the standard one, but seeing what it does to the API I'm reconsidering it.  This forces each logger to repeat a bunch of logic that really is specific to ValidationCallback and unrelated to logging.  It might be better to just make `save_checkpoint()` public and call it whenever we write a checkpoint normally.  What do you think?

Reply:
I wasn't sure at the beginning either, but after working through both approaches I think having the integrated checkpointing (our original approach, the one you described above) seems to be a bit more clean from a code perspective. If we keep the independent checkpointing approach (where save_checkpoint is called in log_batch, log_values) we would have to pass in a lot of checkpointing parameters as you probably saw above (checkpoint_metric, checkpoint_metric_value, checkpoint_on_min) to the logging functions. 

We could expose `save_checkpoint()` as a Logger API method and it would be called anywhere whereever checkpointing is done. The three places I've identified where checkpointing is called are:
1. Model class fit_generator(): https://github.com/kshen3778/deepchem/blob/wandb2/deepchem/models/keras_model.py#L469
2. Model class end of fit: https://github.com/kshen3778/deepchem/blob/wandb2/deepchem/models/keras_model.py#L498
3. ValidationCallback: https://github.com/kshen3778/deepchem/blob/wandb2/deepchem/models/callbacks.py#L96
Let me know if I'm missing any. 
So I'm guessing is that in each of these locations there would be:
```
for ext_logger in self.loggers:
   ext_logger.save_checkpoint(.....)
```
I'm still unsure what the parameters for `save_checkpoint()` would be. I'll add another post once I give it some more thought.

I also don't want to touch the [pre-existing](https://github.com/kshen3778/deepchem/blob/wandb2/deepchem/models/keras_model.py#L1072) save_checkpoint() that exists in the Model classes because they are specific to the default Tensorflow/Pytorch checkpointing behaviour.


On a sort of related note, the one annoying challenge I came across was the checkpoint naming. We want checkpoints coming from different locations (training, eval callback 1, eval callback 2, etc) to have different names when being saved. Right now in Wandb, for each "location" we have all of the location's checkpoints' being saved under `[location].[metric-name].checkpoints_[run_name]`. 

![Screenshot from 2021-08-05 09-59-57](https://user-images.githubusercontent.com/9289959/128363047-84492f97-4739-409a-9e84-dd68a8cbd8eb.png)
![Screenshot from 2021-08-05 10-01-19](https://user-images.githubusercontent.com/9289959/128363125-3880de14-a21c-4c15-8dca-e09dd434ae4e.png)

This is one reason I added a `name` parameter to ValidationCallback in order to distinguish between different callbacks. The name would be [passed into](https://github.com/kshen3778/deepchem/blob/wandb2/deepchem/models/callbacks.py#L110) `location` parameter when logging (and saving checkpoints). The annoying part comes when the user doesn't provide a name for the ValidationCallback in which case we have to [assign it a unique name](https://github.com/kshen3778/deepchem/blob/wandb2/deepchem/models/callbacks.py#L102). I feel like this isn't ideal but I'm not sure if there's a better way. In theory we could name checkpoints under their model_dir but that wouldn't work if mutliple callbacks are sharing a save directory.

Reply:
@peastman I've written up my thoughts above. Anything else you want to add for this?

Reply:
I think I agree.  Let's just make `save_checkpoint()` public and call it anywhere that checkpoints get written.

The "location" argument seems a bit too implementation specific, relating to how it's being saved.  Am I correct that this really just refers to what information is being saved: training loss, validation loss (perhaps for more than one validation set, though that's unusual), etc.?  The ideal would be to pass in semantically meaningful information, and have WandbLogger select a location based on that.

Reply:
Sounds good for `save_checkpoint()`

Regarding the "location" argument: It doesn't necessarily refers to the information being saved. It's more related to the organization system used by the logger dashboard. See this [run](https://wandb.ai/kwshen/deepchem/runs/29oukj1f?workspace=user-kwshen) for example:
![Screenshot from 2021-08-16 15-30-20](https://user-images.githubusercontent.com/9289959/129619569-4be532d3-1ed7-486e-9f3f-9f1bb39c51d0.png)
![Screenshot from 2021-08-16 15-30-31](https://user-images.githubusercontent.com/9289959/129619581-bdf97e28-4e5e-40cc-b228-e407ed1da005.png)
Basically the locations in this case are eval/callback1, eval/callback2, and train. This is because there are two ValidationCallbacks running as well as the training (the two callbacks might be operating on different datasets - for example "val" and "test" or two validation sets).  This structure allows people to understand exactly where their values are coming from. Otherwise, you could very possibly have pearson_r2_score being logged in multiple places (for example in two callbacks, or once in training and once in one the callbacks) and you wouldn't be able to distinguish where it's coming from.

I think it's not just W&B that has this hierarchical location structure. I saw that Neptune, for example also uses this style (https://docs.neptune.ai/you-should-know/logging-metadata). I'm sure other loggers probably follow similar formats.